File: core/src/main/kotlin/br/com/cstag/core/services/analyze/utils/SplitKm.kt
package br.com.cstag.core.services.analyze.utils

import java.math.BigDecimal

data class SplitKm(
    val parts: List<Int>,
    val operator: String?,
    val source: String,
) {
    companion object {
        fun valueOf(km: String): SplitKm {
            var splitOperator: String? = null
            when {
                km.contains("+") -> splitOperator = "+"
                km.contains(".") -> splitOperator = "."
                km.contains(",") -> splitOperator = ","
            }
            val kmParts = if (splitOperator != null) km.split(splitOperator).map { it.trim().toInt() }
                else listOf(km.trim().toInt())
            return SplitKm(kmParts, splitOperator, km)
        }
    }

    fun toBigDecimal(): BigDecimal {
        return BigDecimal(this.parts.joinToString(separator = "."))
    }
}

fun Collection<SplitKm>.sort(): Collection<SplitKm> {
    return this.sortedWith(compareBy { it.toBigDecimal() })
}

File: apps/event-consumer/src/ticket-analysis/split-km.js
import { KMNotFoundException } from './exceptions.js';

export default class SplitKM {
  /**
   * @param {number[]} parts
   * @param {string?} operator
   * @param {string} source
   */
  constructor(parts, operator, source) {
    this.parts = parts;
    this.operator = operator;
    this.source = source;
  }

  /** @param {string} km */
  static valueOf(km) {
    let splitOperator = null;
    if (km.includes('+')) {
      splitOperator = '+';
    } else if (km.includes('.')) {
      splitOperator = '.';
    } else if (km.includes(',')) {
      splitOperator = ',';
    }
    const kmParts =
      splitOperator === null
        ? [Number(km.trim())]
        : km.split(splitOperator).map((i) => Number(i.trim()));

    if (Number.isNaN(kmParts[0])) {
      throw new KMNotFoundException(`Error ao criar KMSplit ${km}`);
    }

    return new SplitKM(kmParts, splitOperator, km);
  }

  toNumber() {
    return Number(this.parts.join('.'));
  }
}
